build_tools/cmake/iree_llvm.cmake

```diff
@@ -0,0 +1,201 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Support for configuring LLVM/MLIR and dependent sub-projects.
+# There are three top-level entry-points:
+#   iree_llvm_configure_bundled() : Configures the bundled LLVM and
+#     sub-projects from third_party.
+#   iree_llvm_configure_installed() : Configures from installed LLVM and
+#     sub-projects.
+#   iree_add_llvm_external_project(...) : Adds an external LLVM project in
+#     a similar fashion to LLVM_EXTERNAL_PROJECTS define (but in a way that
+#     does not require a source build of LLVM).
+
+macro(iree_llvm_configure_bundled)
+  message(STATUS "Adding bundled LLVM source dependency")
+  iree_llvm_set_bundled_cmake_options()
+
+  # Enable MLIR Python bindings if IREE Python bindings enabled.
+  if(IREE_BUILD_PYTHON_BINDINGS)
+    set(MLIR_ENABLE_BINDINGS_PYTHON ON)
+    set(MHLO_ENABLE_BINDINGS_PYTHON ON)
+  endif()
+
+  # Disable LLVM's warnings.
+  set(LLVM_ENABLE_WARNINGS OFF)
+
+  # Stash cmake build type in case LLVM messes with it.
+  set(_CMAKE_BUILD_TYPE "${CMAKE_BUILD_TYPE}")
+
+  add_subdirectory("third_party/llvm-project/llvm" "llvm-project" EXCLUDE_FROM_ALL)
+
+  # Reset CMAKE_BUILD_TYPE to its previous setting.
+  set(CMAKE_BUILD_TYPE "${_CMAKE_BUILD_TYPE}" )
+
+  # Set some CMake variables that mirror things exported in the find_package
+  # world. Source of truth for these is in an installed LLVMConfig.cmake,
+  # MLIRConfig.cmake, LLDConfig.cmake (etc) and in the various standalone
+  # build segments of each project's top-level CMakeLists.
+  set(LLVM_CMAKE_DIR "${IREE_BINARY_DIR}/llvm-project/lib/cmake/llvm")
+  list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
+  # TODO: Fix MLIR upstream so it doesn't spew into the containing project
+  # binary dir.
+  set(MLIR_CMAKE_DIR "${IREE_BINARY_DIR}/lib/cmake/mlir")
+  list(APPEND CMAKE_MODULE_PATH "${MLIR_CMAKE_DIR}")
+
+  set(LLVM_INCLUDE_DIRS
+    ${IREE_SOURCE_DIR}/third_party/llvm-project/llvm/include
+    ${IREE_BINARY_DIR}/llvm-project/include
```

Comment:
Late to the party, but I find it a bit weird that this isn't still nested under `third_party/` like all the other third-party dependencies. LLVM is special though, so it's not that big a deal

Reply:
Yeah, I've been meaning to change this for ~years because of the frequency of reaching into the llvm build dir and reducing keystrokes. It's worth being weird since more efficient to interact with.